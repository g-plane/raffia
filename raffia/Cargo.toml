[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
