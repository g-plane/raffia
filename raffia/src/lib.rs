//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
