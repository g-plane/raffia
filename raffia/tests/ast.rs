use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
