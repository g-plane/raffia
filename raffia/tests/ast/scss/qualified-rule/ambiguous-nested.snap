---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    QualifiedRule(
      type: "QualifiedRule",
      selector: SelectorList(
        type: "SelectorList",
        selectors: [
          ComplexSelector(
            type: "ComplexSelector",
            children: [
              CompoundSelector(
                type: "CompoundSelector",
                children: [
                  TagNameSelector(
                    type: "TagNameSelector",
                    name: WqName(
                      type: "WqName",
                      name: Ident(
                        type: "Ident",
                        name: "foo",
                        raw: "foo",
                        span: Span(
                          start: 0,
                          end: 3,
                        ),
                      ),
                      prefix: None,
                      span: Span(
                        start: 0,
                        end: 3,
                      ),
                    ),
                    span: Span(
                      start: 0,
                      end: 3,
                    ),
                  ),
                ],
                span: Span(
                  start: 0,
                  end: 3,
                ),
              ),
            ],
            span: Span(
              start: 0,
              end: 3,
            ),
          ),
        ],
        commaSpans: [],
        span: Span(
          start: 0,
          end: 3,
        ),
      ),
      block: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          QualifiedRule(
            type: "QualifiedRule",
            selector: SelectorList(
              type: "SelectorList",
              selectors: [
                ComplexSelector(
                  type: "ComplexSelector",
                  children: [
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "bar",
                              raw: "bar",
                              span: Span(
                                start: 8,
                                end: 11,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 8,
                              end: 11,
                            ),
                          ),
                          span: Span(
                            start: 8,
                            end: 11,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "baz",
                            raw: "baz",
                            span: Span(
                              start: 12,
                              end: 15,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 11,
                            end: 15,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "bang",
                            raw: "bang",
                            span: Span(
                              start: 16,
                              end: 20,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 15,
                            end: 20,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "bop",
                            raw: "bop",
                            span: Span(
                              start: 21,
                              end: 24,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 20,
                            end: 24,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "biddle",
                            raw: "biddle",
                            span: Span(
                              start: 25,
                              end: 31,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 24,
                            end: 31,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "woo",
                            raw: "woo",
                            span: Span(
                              start: 32,
                              end: 35,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 31,
                            end: 35,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "look",
                            raw: "look",
                            span: Span(
                              start: 36,
                              end: 40,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 35,
                            end: 40,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "at",
                            raw: "at",
                            span: Span(
                              start: 41,
                              end: 43,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 40,
                            end: 43,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "all",
                            raw: "all",
                            span: Span(
                              start: 44,
                              end: 47,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 43,
                            end: 47,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "these",
                            raw: "these",
                            span: Span(
                              start: 48,
                              end: 53,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 47,
                            end: 53,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "pseudoclasses",
                            raw: "pseudoclasses",
                            span: Span(
                              start: 54,
                              end: 67,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 53,
                            end: 67,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 8,
                        end: 67,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 8,
                    end: 67,
                  ),
                ),
              ],
              commaSpans: [],
              span: Span(
                start: 8,
                end: 67,
              ),
            ),
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "a",
                    raw: "a",
                    span: Span(
                      start: 69,
                      end: 70,
                    ),
                  ),
                  colonSpan: Span(
                    start: 70,
                    end: 71,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "b",
                      raw: "b",
                      span: Span(
                        start: 72,
                        end: 73,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 69,
                    end: 73,
                  ),
                ),
              ],
              span: Span(
                start: 68,
                end: 74,
              ),
            ),
            span: Span(
              start: 8,
              end: 74,
            ),
          ),
          QualifiedRule(
            type: "QualifiedRule",
            selector: SelectorList(
              type: "SelectorList",
              selectors: [
                ComplexSelector(
                  type: "ComplexSelector",
                  children: [
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "bar",
                              raw: "bar",
                              span: Span(
                                start: 78,
                                end: 81,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 78,
                              end: 81,
                            ),
                          ),
                          span: Span(
                            start: 78,
                            end: 81,
                          ),
                        ),
                        PseudoClassSelector(
                          type: "PseudoClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "baz",
                            raw: "baz",
                            span: Span(
                              start: 82,
                              end: 85,
                            ),
                          ),
                          arg: None,
                          span: Span(
                            start: 81,
                            end: 85,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 78,
                        end: 85,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 85,
                        end: 86,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "bang",
                              raw: "bang",
                              span: Span(
                                start: 86,
                                end: 90,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 86,
                              end: 90,
                            ),
                          ),
                          span: Span(
                            start: 86,
                            end: 90,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 86,
                        end: 90,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 90,
                        end: 91,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "bop",
                              raw: "bop",
                              span: Span(
                                start: 91,
                                end: 94,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 91,
                              end: 94,
                            ),
                          ),
                          span: Span(
                            start: 91,
                            end: 94,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 91,
                        end: 94,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 94,
                        end: 95,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "biddle",
                              raw: "biddle",
                              span: Span(
                                start: 95,
                                end: 101,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 95,
                              end: 101,
                            ),
                          ),
                          span: Span(
                            start: 95,
                            end: 101,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 95,
                        end: 101,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 101,
                        end: 102,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "woo",
                              raw: "woo",
                              span: Span(
                                start: 102,
                                end: 105,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 102,
                              end: 105,
                            ),
                          ),
                          span: Span(
                            start: 102,
                            end: 105,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 102,
                        end: 105,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 105,
                        end: 106,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "look",
                              raw: "look",
                              span: Span(
                                start: 106,
                                end: 110,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 106,
                              end: 110,
                            ),
                          ),
                          span: Span(
                            start: 106,
                            end: 110,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 106,
                        end: 110,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 110,
                        end: 111,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "at",
                              raw: "at",
                              span: Span(
                                start: 111,
                                end: 113,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 111,
                              end: 113,
                            ),
                          ),
                          span: Span(
                            start: 111,
                            end: 113,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 111,
                        end: 113,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 113,
                        end: 114,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "all",
                              raw: "all",
                              span: Span(
                                start: 114,
                                end: 117,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 114,
                              end: 117,
                            ),
                          ),
                          span: Span(
                            start: 114,
                            end: 117,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 114,
                        end: 117,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 117,
                        end: 118,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "these",
                              raw: "these",
                              span: Span(
                                start: 118,
                                end: 123,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 118,
                              end: 123,
                            ),
                          ),
                          span: Span(
                            start: 118,
                            end: 123,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 118,
                        end: 123,
                      ),
                    ),
                    Combinator(
                      type: "Combinator",
                      kind: Descendant,
                      span: Span(
                        start: 123,
                        end: 124,
                      ),
                    ),
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "elems",
                              raw: "elems",
                              span: Span(
                                start: 124,
                                end: 129,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 124,
                              end: 129,
                            ),
                          ),
                          span: Span(
                            start: 124,
                            end: 129,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 124,
                        end: 129,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 78,
                    end: 129,
                  ),
                ),
              ],
              commaSpans: [],
              span: Span(
                start: 78,
                end: 129,
              ),
            ),
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "a",
                    raw: "a",
                    span: Span(
                      start: 131,
                      end: 132,
                    ),
                  ),
                  colonSpan: Span(
                    start: 132,
                    end: 133,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "b",
                      raw: "b",
                      span: Span(
                        start: 134,
                        end: 135,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 131,
                    end: 135,
                  ),
                ),
              ],
              span: Span(
                start: 130,
                end: 136,
              ),
            ),
            span: Span(
              start: 78,
              end: 136,
            ),
          ),
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "bar",
              raw: "bar",
              span: Span(
                start: 140,
                end: 143,
              ),
            ),
            colonSpan: Span(
              start: 143,
              end: 144,
            ),
            value: [
              Ident(
                type: "Ident",
                name: "baz",
                raw: "baz",
                span: Span(
                  start: 144,
                  end: 147,
                ),
              ),
              Ident(
                type: "Ident",
                name: "bang",
                raw: "bang",
                span: Span(
                  start: 148,
                  end: 152,
                ),
              ),
              Ident(
                type: "Ident",
                name: "bop",
                raw: "bop",
                span: Span(
                  start: 153,
                  end: 156,
                ),
              ),
              Ident(
                type: "Ident",
                name: "biddle",
                raw: "biddle",
                span: Span(
                  start: 157,
                  end: 163,
                ),
              ),
              Ident(
                type: "Ident",
                name: "woo",
                raw: "woo",
                span: Span(
                  start: 164,
                  end: 167,
                ),
              ),
              Ident(
                type: "Ident",
                name: "look",
                raw: "look",
                span: Span(
                  start: 168,
                  end: 172,
                ),
              ),
              Ident(
                type: "Ident",
                name: "at",
                raw: "at",
                span: Span(
                  start: 173,
                  end: 175,
                ),
              ),
              Ident(
                type: "Ident",
                name: "all",
                raw: "all",
                span: Span(
                  start: 176,
                  end: 179,
                ),
              ),
              Ident(
                type: "Ident",
                name: "these",
                raw: "these",
                span: Span(
                  start: 180,
                  end: 185,
                ),
              ),
              Ident(
                type: "Ident",
                name: "elems",
                raw: "elems",
                span: Span(
                  start: 186,
                  end: 191,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 140,
              end: 191,
            ),
          ),
        ],
        span: Span(
          start: 4,
          end: 194,
        ),
      ),
      span: Span(
        start: 0,
        end: 194,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 195,
  ),
)
